nodes/src/nodes/cost_tracker/tracker.py

```
@@ -0,0 +1,184 @@
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+# =============================================================================
+
+"""Thread-safe cost tracker with budget enforcement."""
+
+import json
+import threading
+from typing import Any, Dict, List, Optional
```

Ruff (0.15.7)

[warning] `typing.Dict` is deprecated, use `dict` instead

(UP035)